### DISCUSSION TITLE
How to avoid UDP fragmentation in application

### DISCUSSION THREAD

**Discussion by songzhao831** [OpenDDS/OpenDDS#4082] [Q&A] [answered]
Hello,
We develope the application on the end-point device and run applications on a PTP networking. (app<->NAT networking with gPTP<..><->app)

But we found a the UDP fragmentation packet is modified by Networking Switch device which support PTP protocal, and hardware treat the UDP fragmented package as UDP PTP frame(Source port in payload is UDP PTP port by coincidence) and add timestamp on the package.
03-1f is the PTP port reserve defined in IEEE1588. "42 62 5e c1" is the modified field (PTP timestamp).

0000   8c 16 45 ba ce 39 00 21 85 c5 2b 8f 08 00 45 00
0010   00 4c 79 19 04 00 40 11 79 d2 c0 a8 01 64 c0 a8
0020   01 01 08 00 _01 3f_ 00 38 bd fd 22 62 63 64 65 66
0030   67 68 69 6a 6b 6c 6d 6e 6f 70 _42 62 5e c1_ 75 76
0040   77 61 62 63 64 65 66 67 68 64 2b 62 8c 86 19 b2
0050   3a 6e 5b 84 00 00 10 03 00 00........

So, it cause the commucation failed. 
How to avoid the UDP fragmentation in the application?

Leo Song.

**Comment by jrw972** [comment, marked as answer]
First, if you are using the UDP transport, then please consider migrating to rtps_udp since the UDP transport will be removed in the next major release.

I don't fully understand your issue so please add more detail if I don't answer your question.

There are two possible solutions here.

1. Change your message types so they are smaller.  This isn't always possible or desirable.
2. Set up a maximum message size to force fragmentation.  This is the MTU of your network minus the IP header, UDP header, etc.

For the UDP transport, you need to set the limit in `max_message_size` of dds/DCPS/transport/udp/UdpSendStrategy.h
For RTPS, see the `SedpMaxMessageSize` and `max_message_size` configuration values https://opendds.readthedocs.io/en/latest/devguide/run_time_configuration.html

**Reply by songzhao831** [reply]
Thank you for your reply.

It may couldn't fix my issue with changing the maximum size of message, because UDP fragmentation package doesn't have UDP header which filled by payload raw data, it still have chance to hit UDP PTP which defined in IEEE1588 and be modified by router/switch.

In IEEE1588 v2, if the ingress package is UDP protocol (0x11), don't care fragmentation bit in IP header on some device, then hardware check the UDP des port==0x013F, if Yes, treat it as PTP frame and pad hardware timestamp into the package, this 'PTP' package will be sent to PTP stack.

Let me know if it still don't clear for you.

Leo

**Reply by jrw972** [reply]
I'm still not getting it.

The point of using RTPS with a `max_message_size` is to avoid fragmentation at the IP/UDP level by fragmenting at the RTPS level.
For example, suppose you set `max_message_size` to 100 bytes and send a sample that is 200 bytes.
Before any UDP packet is sent, OpenDDS will split the 200 byte payload into two 100 byte payloads (fragments).
OpenDDS will then send these 100 byte payloads.
On the receiving side, the two packets will be processed at the 200 byte payload reconstructed.
If the 100 byte payloads are below the MTU, then IP fragmentation will not occur.
If the `max_message_size` is chosen conservatively, PTP can prepend/append to these messages and the resulting message will still be under the MTU so that IP fragmentation will not occur.

I have addressed a similar problem when deploying OpenDDS in a cloud environment.  This manifested in two ways.
First, OpenDDS participants were sending packets that were too large to the cloud.  That is, they exceeded the MTU of certain routers to IP fragmentation was applied.  Some routers don't support IP fragmentation, so only the first packet in the fragment stream arrived.  Setting the `max_message_size` fixed this issue.
Second, the transfer from cloud-based OpenDDS participants to other cloud-based participants had similar problems because IP fragmentation didn't work in the virtualized network and the virtualized network added additional overhead.

**Reply by songzhao831** [reply]
I get your point.

There is the package on the Switch/NIC captured by wireshark:
![image](https://user-images.githubusercontent.com/7333870/225888914-4a54a914-c13c-471d-83ce-eab798ae772e.png)

I think the RTPS could fix this issue, if the fragmentation at RTPS level instead of IP/UDP level, it can avoid the Switch/Router to treat it as OSI layer 2 package I believe, but I need the fragmentation packet to confirm it.

Could you capture one fragment packet by wireshark like above pic?

Thanks for your reply.

Leo

**Reply by songzhao831** [reply]
Hi Justin,

It's related to TCP/IP and MAC layer(Router/Switch) data transport. 

If I understand correctly, RTPS UDP protocal adds RTPS header in the UDP header, which could avoid the UDP header field is filled by payload data. If yes,  it's the best solution.

Another question about the UDP transport, you mentioned set UDP_MAX_MESSAGE_SIZE < MTU (1500 ), if the UDP header field still be filled by payload(random data), it still have chance to hit UDP PTP recognization(Dst port=0x013f), is it right?

Leo

**Reply by jrw972** [reply]
The LargeSample test in tests/DCPS/LargeSample uses fragmentation.  Use a packet capture with `./run_test.pl rtps` capture RTPS messages using fragmentation.

OpenDDS always sends individual UDP messages, i.e., it does not send raw IP packets.  Assuming the OS is working, it should prepend it with a proper UDP header and the router/switch will always see a proper UDP packet.

**Comment by songzhao831** [comment]
Thanks, It can be close.
